File: Components/WinForms/CS/DiagramMainDemo/Modules/ProductFlowDiagramModule.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Drawing;
using System.Data;
using System.Linq;
using System.Text;
using System.Windows.Forms;
using DevExpress.XtraDiagram.Demos;
using DevExpress.DXperience.Demos;
using DevExpress.Diagram.Core;
using System.Data.OleDb;
using DevExpress.Diagram.Demos;
using System.Collections.ObjectModel;
using DevExpress.Utils;
using DevExpress.Data.Filtering;

namespace DevExpress.XtraDiagram.Demos {
    public partial class ProductFlowDiagramModule : DiagramTutorialControl {
        readonly DiagramItemStyleId[] styles = DiagramShapeStyleId.Styles.ToArray();
        readonly ProductFlowInfo info;


        public ProductFlowDiagramModule() {
            InitializeComponent();
            info = OrderDataGenerator.GenerateProductFlowInfo();
            diagramDataBindingController1.BeginInit();
            diagramDataBindingController1.DataSource = info.Items;
            diagramDataBindingController1.ConnectorsSource = info.ProductFlows;
            diagramDataBindingController1.EndInit();
            gridControl.DataSource = info.Orders;
            Diagram.SelectionChanged += OnDiagramSelectionChanged;
            Diagram.SelectItem(Diagram.Items.First());
        }
        void OnDiagramSelectionChanged(object sender, DiagramSelectionChangedEventArgs e) {
            var selectedDiagramItem = Diagram.PrimarySelection;
            gridView.ClearGrouping();
            gridView.ActiveFilterCriteria = null;

            if(selectedDiagramItem == null)
                return;

            var customer = selectedDiagramItem.DataContext as CustomerData;
            if(customer != null) {
                gridView.ActiveFilterCriteria = new BinaryOperator("Customer.Name", customer.Name, BinaryOperatorType.Equal);
                gridView.Columns["Category.Name"].Group();
                gridView.ExpandAllGroups();
            }

            var category = selectedDiagramItem.DataContext as CategoryData;
            if(category != null) {
                gridView.ActiveFilterCriteria = new BinaryOperator("Category.Name", category.Name, BinaryOperatorType.Equal);
                gridView.Columns["Customer.Name"].Group();
                gridView.ExpandAllGroups();
            }

            var connector = selectedDiagramItem.DataContext as ProductFlowData;
            if(connector != null) {
                var productFilter = new BinaryOperator("Category.Name", connector.Category.Name, BinaryOperatorType.Equal);
                var customerFilter = new BinaryOperator("Customer.Name", connector.Customer.Name, BinaryOperatorType.Equal);
                gridView.ActiveFilterCriteria = new GroupOperator(GroupOperatorType.And, productFilter, customerFilter);
            }
        }
        private void diagramDataBindingController1_GenerateItem(object sender, DiagramGenerateItemEventArgs e) {
            var templateName = (e.DataObject is CustomerData) ? "CustomerTemplate" : "CategoryTemplate";
            e.Item = e.CreateItemFromTemplate(templateName);
        }

        private void diagramDataBindingController1_CustomLayoutItems(object sender, DiagramCustomLayoutItemsEventArgs e) {
            ArrangeItemsInLine<CategoryData>(e.Items, new PointFloat(600, 50), new Size(150, 105), 20);
            ArrangeItemsInLine<CustomerData>(e.Items, new PointFloat(50, 100), new Size(150, 105), 20);
            foreach(var item in e.Items) {
                var customer = item.DataContext as CustomerData;
                if(customer != null) {
                    item.ThemeStyleId = styles[Array.IndexOf(info.Customers, customer)];
                }
            }
            foreach(var connector in e.Connectors) {
                var connectorData = (ProductFlowData)connector.DataContext;
                connector.ThemeStyleId = styles[Array.IndexOf(info.Customers, connectorData.Customer)];
            }
            e.Handled = true;
        }
        void ArrangeItemsInLine<TDataContext>(IEnumerable<DiagramItem> items, PointFloat startPosition, Size itemSize, int margin) {
            PointFloat position = startPosition;
            foreach(var diagramItem in items.Where(x => x.DataContext is TDataContext)) {
                diagramItem.Position = position;
                position.Offset(0, itemSize.Height + margin);
            }
        }

        private void diagramDataBindingController1_UpdateConnector(object sender, DiagramUpdateConnectorEventArgs e) {
            var connectorData = (ProductFlowData)e.DataObject;
            e.Connector.Appearance.BorderSize = (int)connectorData.Weight;
        }
    }
}

File: Components/WinForms/VB/DiagramMainDemo/Modules/ProductFlowDiagramModule.vb
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Drawing
Imports System.Data
Imports System.Linq
Imports System.Text
Imports System.Windows.Forms
Imports DevExpress.XtraDiagram.Demos
Imports DevExpress.DXperience.Demos
Imports DevExpress.Diagram.Core
Imports System.Data.OleDb
Imports DevExpress.Diagram.Demos
Imports System.Collections.ObjectModel
Imports DevExpress.Utils
Imports DevExpress.Data.Filtering

Namespace DevExpress.XtraDiagram.Demos
    Partial Public Class ProductFlowDiagramModule
        Inherits DiagramTutorialControl

        Private ReadOnly styles() As DiagramItemStyleId = DiagramShapeStyleId.Styles.ToArray()
        Private ReadOnly info As ProductFlowInfo


        Public Sub New()
            InitializeComponent()
            info = OrderDataGenerator.GenerateProductFlowInfo()
            diagramDataBindingController1.BeginInit()
            diagramDataBindingController1.DataSource = info.Items
            diagramDataBindingController1.ConnectorsSource = info.ProductFlows
            diagramDataBindingController1.EndInit()
            gridControl.DataSource = info.Orders
            AddHandler Diagram.SelectionChanged, AddressOf OnDiagramSelectionChanged
            Diagram.SelectItem(Diagram.Items.First())
        End Sub
        Private Sub OnDiagramSelectionChanged(ByVal sender As Object, ByVal e As DiagramSelectionChangedEventArgs)
            Dim selectedDiagramItem = Diagram.PrimarySelection
            gridView.ClearGrouping()
            gridView.ActiveFilterCriteria = Nothing

            If selectedDiagramItem Is Nothing Then
                Return
            End If

            Dim customer = TryCast(selectedDiagramItem.DataContext, CustomerData)
            If customer IsNot Nothing Then
                gridView.ActiveFilterCriteria = New BinaryOperator("Customer.Name", customer.Name, BinaryOperatorType.Equal)
                gridView.Columns("Category.Name").Group()
                gridView.ExpandAllGroups()
            End If

            Dim category = TryCast(selectedDiagramItem.DataContext, CategoryData)
            If category IsNot Nothing Then
                gridView.ActiveFilterCriteria = New BinaryOperator("Category.Name", category.Name, BinaryOperatorType.Equal)
                gridView.Columns("Customer.Name").Group()
                gridView.ExpandAllGroups()
            End If

            Dim connector = TryCast(selectedDiagramItem.DataContext, ProductFlowData)
            If connector IsNot Nothing Then
                Dim productFilter = New BinaryOperator("Category.Name", connector.Category.Name, BinaryOperatorType.Equal)
                Dim customerFilter = New BinaryOperator("Customer.Name", connector.Customer.Name, BinaryOperatorType.Equal)
                gridView.ActiveFilterCriteria = New GroupOperator(GroupOperatorType.And, productFilter, customerFilter)
            End If
        End Sub
        Private Sub diagramDataBindingController1_GenerateItem(ByVal sender As Object, ByVal e As DiagramGenerateItemEventArgs) Handles diagramDataBindingController1.GenerateItem
            Dim templateName = If(TypeOf e.DataObject Is CustomerData, "CustomerTemplate", "CategoryTemplate")
            e.Item = e.CreateItemFromTemplate(templateName)
        End Sub

        Private Sub diagramDataBindingController1_CustomLayoutItems(ByVal sender As Object, ByVal e As DiagramCustomLayoutItemsEventArgs) Handles diagramDataBindingController1.CustomLayoutItems
            ArrangeItemsInLine(Of CategoryData)(e.Items, New PointFloat(600, 50), New Size(150, 105), 20)
            ArrangeItemsInLine(Of CustomerData)(e.Items, New PointFloat(50, 100), New Size(150, 105), 20)
            For Each item In e.Items
                Dim customer = TryCast(item.DataContext, CustomerData)
                If customer IsNot Nothing Then
                    item.ThemeStyleId = styles(Array.IndexOf(info.Customers, customer))
                End If
            Next item
            For Each connector In e.Connectors
                Dim connectorData = CType(connector.DataContext, ProductFlowData)
                connector.ThemeStyleId = styles(Array.IndexOf(info.Customers, connectorData.Customer))
            Next connector
            e.Handled = True
        End Sub
        Private Sub ArrangeItemsInLine(Of TDataContext)(ByVal items As IEnumerable(Of DiagramItem), ByVal startPosition As PointFloat, ByVal itemSize As Size, ByVal margin As Integer)
            Dim position As PointFloat = startPosition
            For Each diagramItem In items.Where(Function(x) TypeOf x.DataContext Is TDataContext)
                diagramItem.Position = position
                position.Offset(0, itemSize.Height + margin)
            Next diagramItem
        End Sub

        Private Sub diagramDataBindingController1_UpdateConnector(ByVal sender As Object, ByVal e As DiagramUpdateConnectorEventArgs) Handles diagramDataBindingController1.UpdateConnector
            Dim connectorData = CType(e.DataObject, ProductFlowData)
            e.Connector.Appearance.BorderSize = CInt((connectorData.Weight))
        End Sub
    End Class
End Namespace
